src/event-log.c

```diff
@@ -0,0 +1,402 @@
+ *
+ * @return newly-allocated formatted string (without trailing newline)
+ */
+static gchar *event_log_format_fields_readable(GLogLevelFlags log_level,
+		const GLogField *fields, gsize n_fields, gboolean verbose)
+{
+	g_autoptr(GDateTime) now = NULL;
+	g_autofree gchar *now_formatted = NULL;
+	GString *gstring;
+
+	gstring = g_string_new(NULL);
+	now = g_date_time_new_now_utc();
+	now_formatted = g_date_time_format(now, "%Y-%m-%dT%H:%M:%SZ");
+	g_string_append_printf(gstring, "%s: ", now_formatted);
+
+	for (gsize j = 0; j < n_fields; j++) {
+		const GLogField *ifield = &fields[j];
+		const gchar *message = NULL;
+
+		/* print message */
+		if (g_strcmp0(ifield->key, "MESSAGE") == 0) {
+			message = ifield->value;
+			g_string_append_printf(gstring, "%s", message);
+			break;
+		}
+	}
+
+	/* skip other fields if non-verbose */
+	if (!verbose)
+		return g_string_free(gstring, FALSE);
+
+	/* collect verbose fields in order of appearance */
+	for (gsize j = 0; j < n_fields; j++) {
+		const GLogField *ifield = &fields[j];
+		const gchar *key = NULL;
+		const gchar *value = NULL;
+
+		if (g_strcmp0(ifield->key, "TRANSACTION_ID") == 0) {
+			key = "transaction ID";
+			value = ifield->value;
+		} else if (g_strcmp0(ifield->key, "BUNDLE_HASH") == 0) {
+			key = "bundle hash";
+			value = ifield->value;
+		} else if (g_strcmp0(ifield->key, "BOOT_ID") == 0) {
+			key = "boot ID";
+			value = ifield->value;
+		}
+
+		if (key) {
+			g_string_append_printf(gstring, "\n                      %s: %s", key, value);
+		}
+	}
+
+	return g_string_free(gstring, FALSE);
+}
+
+/**
+ * Rotates log files.
+ *
+ * If a maximum size is configured and the loggers current size + provided
+ * 'next_len' exceed this maximum size, the method will close the current
+ * output stream, and copy the file to a .1 copy.
+ *
+ * All previously existing rotation files will be rotated by +1.
+ *
+ * If max-files is set, rotation is limited to max-files files.
+ *
+ * @param logger Logger to rotate file for
+ * @param next_len Length of next data to add.
+ *        This is used to determine if the log file needs to be rotated.
+ */
+static void rotate_logfiles(REventLogger* logger, goffset next_len)
+{
+	g_autoptr(GFile) logfile = NULL;
+	g_autofree gchar *rotfile = NULL;
+	g_autoptr(GError) ierror = NULL;
+	guint existing_files = 0;
+	guint files_to_rotate;
+
+	g_return_if_fail(logger);
+
+	/* method is no-op if no maxsize is configured */
+	if (!logger->maxsize)
+		return;
+
+	/* return if size limit not reached yet */
+	if (logger->filesize + next_len <= logger->maxsize)
+		return;
+
+	/* Close open stream */
+	g_clear_pointer(&logger->logstream, g_object_unref);
+
+	/* determine number of existing rotation files */
+	for (guint num = 0; num < G_MAXUINT; num++) {
+		g_autofree gchar *tryfile = g_strdup_printf("%s.%d", logger->filename, num);
+
+		if (g_file_test(tryfile, G_FILE_TEST_EXISTS)) {
+			existing_files++;
+		} else {
+			break;
+		}
+	}
```

Comment:
Can this be dropped by always trying to rotate `maxfiles` files and ignoring `ENOENT` on `rename`?

Reply:
I've thought about this, too. But have not found a proper solution for not having maxfiles set. Or there must be a default maxfiles of like 100 files or so.

Reply:
I'd say a default of 10 (so 1…9) would be fine. That would keep the size fluctuation low enough.

Reply:
10 would be fine for me. Currently maxfiles is defined as the number of rotation files to keep (thus would be 1..10). Do you think this should be changed?

Reply:
I would have included the currently active file in the count. So maxfiles would not be exceeded in total.

Reply:
As an alternative: In `logrotate` the option is called `rotate` and a `rotate` of `0` means "old versions are removed rather than rotated". But `-1` there means "old logs are not removed at all" (the ability we just removed) so we probably shouldn't adapt to its naming anyway.

For `maxfiles` it does not feel that clear how it can be interpreted. But maybe counting this as the total number of files (as you suggested) is fine.

Reply:
Changed the code now to interpret `max-files` as the maximum of all files (including the logfile itself).